KEYMAP_*+_GENERIC macros and keymasks

noseglasses opened the pull request:
I am working on a plugin that provides extended keymap features:
https://github.com/CapeLeidokos/Kaleidoscope-XKeymaps

It's almost finished but to make it work, I need all keyboard hardware headers to define a macro that works similar to the existing `KEYMAP_STACKED` macro. The only difference is that the new macro generates a linear list of entities instead of a braced list. The new macro has an additional first macro parameter `dflt` (=default, name shortened to match row/col macro arguments as r0c0) that is used where the `KEYMAP_STACKED` macro would specify `XXX` for unused key matrix positions.

I rely on these macros with `Kaleidoscope-XKeymaps` to conveniently initialize keymap bitfields (keymasks) that are stored in PROGMEM (see the usage of `KEYMASK_STACKED_PROGMEM` in the example provided in the README.md of `Kaleidoscope-XKeymaps`). As a sidenote, the whole bitfield for the Model01 is really stored in 8 PROGMEM bytes and automatically initialized at compile time.

The newly introduced macros will possibly be useful in other places whenever information is supposed to be assigned to keys and this information is meant to be specified in a `KEYMAP_STACKED`-like fashion but passed as a linear list of entities, e.g. as arguments to a function or as an initialization list.

algernon commented:
This makes sense, though I wonder if it wouldn't be better to replace the current `KEYMAP_STACKED` / `KEYMAP` macros with something similar, instead of adding a new representation?

(Also, can you add a `Signed-Off-By:` line to your commits?)

noseglasses commented:
The `KEYMAP_STACKED` macros create lists that are brace initialized. They are used to initialize 2D arrays. 

The newly introduced `KEYMAP_STACKED_LIST` macros generate linear lists without braces. Those are e.g. useful to pass information that is defined in the sketch in a stacked-keymap fashion as parameters to functions or other macros. There the braced `KEYMAP_STACKED` cannot be used.

The `KEYMAP_STACKED` macros could be replaced by the newly introduced macros only if the two dimensional keymap array would be replaced by a 1D array. That's something that has been suggested in other places before, e.g. [here](https://github.com/keyboardio/Kaleidoscope/pull/589#pullrequestreview-208709849). I would also favor such a design for multiple reasons, e.g. key address handling.

But as that is a different matter I introduced the `KEYMAP_STACKED_LIST` for the time being. If there would one day be a decision to switch over to 1D arrays, `KEYMAP_STACKED` could become a simple wrapper in `Kaleidoscope.h` that wraps specialized versions of `KEYMAP_STACKED_LIST` that are defined by the hardware headers.

P.S.: I tried the signoff stuff and force pushed the signed version but with no luck (see the listed force push above). I will that again, though.

noseglasses commented:
Took me a while to find out what was wrong. All commits must of course be signed off. One was, the other wasn't. Github does not show which commit was the culprit. Squashed and force pushed.

obra commented:
Sorry, my comment from earlier in the week didn't go through.

I really, really don't like requiring device implementors to create additional versions of these macros.  Is there anything you can do to not require every hardware implementation to have duplicated "list" versions of the keymap macros?

noseglasses commented:
> I really, really don't like requiring device implementors to create additional versions of these macros.

Very much understandable.

I know a lot of macro black magic but magical power is not strong enough to which away those curly braces created by the current `KEYMAP_STACKED` macro. Macro tricks can remove ordinary brackets but curly braces are untouchable.

If it was me to decide, I would throw away the current implementation of `KEYMAP_STACKED` immediately because of its lack of flexibility compared to the proposed linear approach. This would require moving from two dimensional `keymap` arrays to one dimensional ones. This is no big deal as those are only used internally by the `Layer` class and everything is wrapped by macros already, thus the transition would be pretty painless. 

As it cannot be 100% excluded that someone relied on the current 2D implementation in a plugin, i propose a [wrapper approach](https://godbolt.org/z/fcS35m) with deprecation for a transition time. As it turns out, the wrapper access, the direct access to the linear array and the access to 2D arrays requires pretty much the same number of instructions.

noseglasses commented:
To make it easier to compare the performance of access to a linear array compared to a two-dimensional array, I [slightly modified](https://godbolt.org/z/9KmOIP) the previous example and wrapped the array access with individual functions.

I simply counted the number of instructions of the individual functions in the example (`oldStyleAccess`, `wrapperAccess`, `indexedLinearAccess`, `indexedLinearAccessThroughInlinedAccessFunction`). The number of instructions required for the different modes of access are as follows. 

Old style indexed row/col access: __26__
Wrapper access with same signature (for transition period): __29__
Direct access to linear array: __27__
Access to linear array through inlined convenience function: __27__ (preferred solution)

This means that eventually we will have only one instruction per access more by going from 2D to 1D keymap arrays but gain much more widely usable `KEYMAP...` macros.

FWIW, the access through a convenience function would e.g. read `keyFromPROGMEMKeymap(layer, row, col)`.

I'd be happy to submit a PR for the overall change (2D->1D + modified `KEYMAP_STACKED` and `KEYMAP` macros) that would be entirely backward compatible. This would completely avoid the `KEYMAP_STACKED_LIST` macros proposed by this PR.

noseglasses commented:
@obra, could you please comment on my ideas about a transition to 1D keymap arrays. The reason why i bother you is that the transition of the `KEYMAP_STACKED` macros is the only missing peace that prevents me right now from publishing my [Kaleidoscope-XKeymaps](https://github.com/CapeLeidokos/Kaleidoscope-XKeymaps) plugin.

Also the keymask feature proposed [here](https://github.com/keyboardio/Kaleidoscope/issues/604) would rely on 1D `KEYMAP...` macros.

I would be happy to submit a PR that demonstrates the necessary changes in the Kaleidoscope core.

obra commented:
@algernon and I have discussed it and agree that moving to a 1D array of keys is probably the right thing for kaleidoscope globally going forward. 

> On Mar 20, 2019, at 4:05 AM, noseglasses <notifications@github.com> wrote:
> 
> @obra, could you please comment on my ideas about a transition to 1D keymap arrays. The reason why i bother you is that the transition of the KEYMAP_STACKED macros is the only missing peace that prevents me right now from publishing my Kaleidoscope-XKeymaps plugin.
> 
> Also the keymask feature proposed here would rely on 1D KEYMAP... macros.
> 
> I would be happy to submit a PR that demonstrates the necessary changes in the Kaleidoscope core.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

noseglasses commented:
Alright, thanks, I will submit a PR.

noseglasses commented:
I just force-pushed an update that does away with the additional `LINEAR` versions of keymaps and instead modifies the overall keymap handling to go with linear keymap arrays. As suggested above in this thread, a compatibility wrapper class was introduced that mimics 2D array access.

Thus, this PR is entirely backward compatible.

We even save some bytes of PROGMEM: [kaleidoscope_change_report__linear_keymap_arrays.pdf](https://github.com/keyboardio/Kaleidoscope/files/3008743/kaleidoscope_change_report__linear_keymap_arrays.pdf)

noseglasses commented:
I just added another commit that adds the keymasks that I had announced here: https://github.com/keyboardio/Kaleidoscope/issues/604

Keymasks will be a key feature of my new xkeymaps plugin but I am pretty sure that they will beneficial in other places of the firmware, too.

Technically the second commit is orthogonal to the first one of this PR. But the second one does depend on the modifications to the `KEYMAP` and `KEYMAP_STACKED` macros that are induced by the first commit. Thus, I find it convenient to have them both together. 

I added a detailed description of the keymask features to the commit message of the second commit.

I changed the title of this PR accordingly.

@algernon and @obra, is that ok for you or do you want me to add a separate PR?

noseglasses commented:
There's one thing that i forgot to ask.

I added the files `keymask.h` and `keymask.cpp` to `Kaleidoscope/src/kaleidoscope`. As I know little about Arduinos internal policy about what to compile and what not, i wonder if `keymask.cpp` will always be compiled? On the one hand, that's good, because there are no linker errors of missing symbols to be expected and the linker does eliminate unused symbols anyway. On the other hand we should restrict the amount of compiled code to only such parts that are actually going to be linked into the firmware. Firmware compiles should be as fast as possible to make users and developers (smoke-examples) happy.

i am a bit lost here.

gedankenexperimenter commented:
From experience, I can tell you that arduino-builder compiles all `*.cpp` files it finds in the source tree of any libraries that get included via the header files referred to in the sketch. Just try putting an unrelated `.cpp` file with a syntax error somewhere, and you'll see.

noseglasses commented:
> From experience, I can tell you that arduino-builder compiles all *.cpp files it finds in the source tree of any libraries that get included via the header files referred to in the sketch. Just try putting an unrelated .cpp file with a syntax error somewhere, and you'll see.

Ok, thanks for the clarification. That's what i suspected. In the case of the `Bitfield` stuff it's fortunately just four tiny functions.

obra commented on src/kaleidoscope/keymask.cpp:
```diff
@@ -0,0 +1,53 @@
+/* Kaleidoscope-XKeymaps
```
Looks like this was extracted from elsewhere..

obra commented on src/kaleidoscope/keymask.h:
```diff
@@ -0,0 +1,202 @@
+/* Kaleidoscope-XKeymaps
```
extracted from elsewhere?

obra commented:
I don't -think- I like the term 'keymask'  - since it's a keymap bitfield, maybe we should just call it a keymap bitfield?

You mention "As it cannot be 100% excluded that someone relied on the current 2D implementation in a plugin, i propose a wrapper approach with deprecation for a transition time. As it turns out, the wrapper access, the direct access to the linear array and the access to 2D arrays requires pretty much the same number of instructions." 

I'm ok with wrappers and deprecation, but I'm having trouble thinking of where this -could- be used from a Plugin.

noseglasses commented:
> but I'm having trouble thinking of where this -could- be used from a Plugin.

Any plugin could theoretically contain code that does `pgm_read_word(&keymap[layer][row][col])`. Although I agree that it is not very likely.

noseglasses commented:
I updated the second commit
* files `keymask.*` are now named `bitfields.*`
* the file headers have been fixed for Kaleidoscope
* the macros `KEYMASK` and `KEYMASK_STACKED` have been renamed `KEYMAP_BITFIELD` and `KEYMAP_BITFIELD_STACKED`

noseglasses commented:
Just noticed that the ergodox hardware internally uses a keymap bitifield. This would be a perfect use case for the newly introduced `Bitfield` class.

noseglasses commented:
`ATMegaKeyboard` also uses a self made bitfield that could be wrapped/replaced.

noseglasses commented:
@algernon, @obra, could you please have a look at this PR. I really would like to release my XKeymaps-plugin which unfortunately depends on this.

obra commented:
I've asked @algernon to dig into this ASAP
ᐧ

On Wed, May 1, 2019 at 12:26 AM noseglasses <notifications@github.com>
wrote:

> @algernon <https://github.com/algernon>, @obra <https://github.com/obra>,
> could you please have a look at this PR. I really would like to release my
> XKeymaps-plugin which unfortunately depends on this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope/pull/600#issuecomment-488226373>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAALC2CSEEU2BOY3X6NO4RTPTFA2BANCNFSM4G5QNDUQ>
> .
>

obra commented:
I'm really sorry we're late on this. Algernon and I have both been slammed

> ᐧ

obra commented:
I started splitting this up into smaller commits, breaking out the bitfield classes into a commit and the wavepool example cleanup into a commit.

I've pushed both of those to master.

This makes it a little bit easier to deal with the rest of it.

I'm once again trying to figure out if there are any actual usages of the "old-style" keymap lookups outside the core. Ideally, we'd be able to avoid having the "_GENERIC" variants.

noseglasses commented:
The  `KEYMAP_*_GENERIC` macros are useful far beyond the definition of actual keymaps. They can serve to elegantly define general per-key information. For such purposes, it is useful to be able to assign special values to those keys that are not used (holes). When defining a bitfield, one might want to choose if the holes are assigned a one or a zero value.

I'm not sure about the names `KEYMAP_*_GENERIC`, though. The term `KEYMAP` might be missleading. But to let each hardware define such a macro instead of the original `KEYMAP_*` macros provides much greater flexibility.

For my [Kaleidoscope-LEDEffect-Maelstrom](
https://github.com/CapeLeidokos/Kaleidoscope-LEDEffect-Maelstrom) plugin, I use the `KEYMAP_STACKED_GENERIC` macro to specify per-key cartesian coordinates. The outcome looks as follows:

![whirl](https://user-images.githubusercontent.com/29701465/57477754-aaa4ca80-7299-11e9-8047-971218cd331c.gif)

See [here](https://github.com/CapeLeidokos/Kaleidoscope/blob/ng_coordinate_infrastructure/src/kaleidoscope/hardware/keyboardio/Model01.h) for definition of the cartesian key-coordinates (at the bottom of the file).

FWIW, there's no floating point math involved at runtime. Everything happens at compiletime and it costs only 128 bytes of PROGMEM to store some integer representations of per-key polar coordinates.

The [Kaleidoscope-XKeymaps](https://github.com/CapeLeidokos/Kaleidoscope-XKeymaps) plugin will also rely on this PR. it uses key-bitfields that are defined via the `*_GENERIC` macros.

obra commented:
Right. I was trying to figure out if we can just drop the _GENERIC part of the name. There’s gotta be a good descriptive name. FWIW, that’s my only current concern about the PR. Naming is hard. 

> On May 9, 2019, at 11:53 AM, noseglasses <notifications@github.com> wrote:
> 
> The KEYMAP_*_GENERIC macros are useful far beyond the definition of actual keymaps. They can serve to elegantly define general per-key information. For such purposes, it is useful to be able to assign special values to those keys that are not used (holes). When defining a bitfield, one might want to choose if the holes are assigned a one or a zero value.
> 
> I'm not sure about the names KEYMAP_*_GENERIC, though. The term KEYMAP might be missleading. But to let each hardware define such a macro instead of the original KEYMAP_* macros provides much greater flexibility.
> 
> For my Kaleidoscope-LEDEffect-Maelstrom plugin, I use the KEYMAP_STACKED_GENERIC macro to specify per-key cartesian coordinates. The outcome looks as follows:
> 
> 
> 
> See here for definition of the cartesian key-coordinates (at the bottom of the file).
> 
> FWIW, there's no floating point math involved at runtime. Everything happens at compiletime and it costs only 128 bytes of PROGMEM to store some integer representations of per-key polar coordinates.
> 
> The Kaleidoscope-XKeymaps plugin will also rely on this PR. it uses key-bitfields that are defined via the *_GENERIC macros.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

noseglasses commented:
Some naming ideas: `KEYWISE_DATA`, `PER_KEY_DATA`, `KEY_PAYLOAD` or, as those macros generate a list, `KEYWISE_DATA_LIST`, etc.

obra commented:
PER_KEY_DATA is what I was going to propose. Given two independent suggestions of that name, let’s run with it. 

> On May 9, 2019, at 12:25 PM, noseglasses <notifications@github.com> wrote:
> 
> Some naming ideas: KEYWISE_DATA, PER_KEY_DATA, KEY_PAYLOAD or, as those macros generate a list, KEYWISE_DATA_LIST, etc.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

obra commented:
Ok. I've made the naming change and rebased on Master. 

https://github.com/keyboardio/Kaleidoscope/commit/547db75474fd48d42df5ae17ec39f1e942086d68 is the commit containing everything as-yet unmerged.

@algernon - I'd love a thumbs up / down on the current state. @noseglasses: have I butchered your work?

noseglasses commented:
@obra, changes look good. I definitely like the new macro name.

algernon commented:
:+1: 

We should add this to `UPGRADING.md` and `NEWS.md`, and decide on a removal date. I think we used about 3 months in the past. How does a removal date of August 13 sound?

obra commented:
great

> On May 10, 2019, at 8:41 AM, Gergely Nagy <notifications@github.com> wrote:
> 
> 👍
> 
> We should add this to UPGRADING.md and NEWS.md, and decide on a removal date. I think we used about 3 months in the past. How does a removal date of August 13 sound?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

obra commented:
It took a while, but this should now be fully merged, including NEWS and
UPGRADING entries. @noseglasses, thanks so much for your patience. Can you
confirm that everything looks ok?
ᐧ

On Fri, May 10, 2019 at 8:41 AM Gergely Nagy <notifications@github.com>
wrote:

> 👍
>
> We should add this to UPGRADING.md and NEWS.md, and decide on a removal
> date. I think we used about 3 months in the past. How does a removal date
> of August 13 sound?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope/pull/600#issuecomment-491334367>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAALC2ABCUK7XFPAQIM374TPUWJUNANCNFSM4G5QNDUQ>
> .
>

noseglasses commented:
@obra, if you are talking about https://github.com/keyboardio/Kaleidoscope/commit/547db75474fd48d42df5ae17ec39f1e942086d68, that looks ok to me. Did you run smoke-examples on that version? 

But i did not quite understand you statement about NEWS and UPGRADING. https://github.com/keyboardio/Kaleidoscope/commit/547db75474fd48d42df5ae17ec39f1e942086d68 doesn't seem to touch these.

obra commented:
The merged head is
https://github.com/keyboardio/Kaleidoscope/commit/b818be1d8ebb376bae86c49082d6405f943dac56

The smoke test from travis with the examples run is
https://travis-ci.org/keyboardio/Kaleidoscope/builds/532989203
ᐧ

On Wed, May 15, 2019 at 1:50 PM noseglasses <notifications@github.com>
wrote:

> @obra <https://github.com/obra>, if you are talking about 547db75
> <https://github.com/keyboardio/Kaleidoscope/commit/547db75474fd48d42df5ae17ec39f1e942086d68>,
> that looks ok to me. Did you run smoke-examples on that version?
>
> But i did not quite understand you statement about NEWS and UPGRADING.
> 547db75
> <https://github.com/keyboardio/Kaleidoscope/commit/547db75474fd48d42df5ae17ec39f1e942086d68>
> doesn't seem to touch these.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope/pull/600?email_source=notifications&email_token=AAALC2A4GF7OLORU4BDY4TTPVRZSJA5CNFSM4G5QNDU2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVP5BYY#issuecomment-492818659>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAALC2HDVLDB3O2ELXJ6MQ3PVRZSJANCNFSM4G5QNDUQ>
> .
>

noseglasses commented:
I see. I got a bit confused due to your use of separate branches. Anyway, thanks a lot for review and merge!

Do you want me to close this PR?

Can the https://github.com/keyboardio/Kaleidoscope/tree/ng_keymap_stacked_list branch be deleted?

noseglasses commented:
BTW, both NEWS and UPGRADING in head look ok.

obra commented:
I've closed it :)

noseglasses commented:
@obra, could you please pull up the bundle repo.

obra commented:
Done. (Though I -believe- you have permissions to do that, too, now :)
ᐧ

On Fri, May 17, 2019 at 4:36 AM noseglasses <notifications@github.com>
wrote:

> @obra <https://github.com/obra>, could you please pull up the bundle repo.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope/pull/600?email_source=notifications&email_token=AAALC2GR44ZSDBYQ5GQ7XWLPV2KFNA5CNFSM4G5QNDU2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVUQQHA#issuecomment-493422620>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAALC2AF4TQ6MCE2UIGRDLDPV2KFNANCNFSM4G5QNDUQ>
> .
>

noseglasses commented:
> Done. (Though I -believe- you have permissions to do that, too, now :)

Permissions yes, but so far no official approval ;-)

Are there any restrictions or is it always ok to pull up any of the bundle repos to the current HEAD?

algernon commented:
It is usually ok to pull up any of the repos to current HEAD, so go ahead.

obra commented:
HEAD of the various repositories -should- always be releasable. In general,
if the subrepository travis builds are coming up green, it's always ok to
"make blindly-commit-updates" in Kaleidoscope-Bundle-Keyboardio.

Best,
Jesse
ᐧ

On Tue, May 21, 2019 at 10:52 PM Gergely Nagy <notifications@github.com>
wrote:

> It is usually ok to pull up any of the repos to current HEAD, so go ahead.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/keyboardio/Kaleidoscope/pull/600?email_source=notifications&email_token=AAALC2EIK2H53TH6BN36QP3PWTNRTA5CNFSM4G5QNDU2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODV57EJQ#issuecomment-494662182>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAALC2AF3WVYZHXAQETIBYLPWTNRTANCNFSM4G5QNDUQ>
> .
>
